src/main/java/org/broadinstitute/hellbender/engine/FeatureWalker.java

```diff
@@ -0,0 +1,108 @@
+        }
+
+        if ( hasIntervals() ) {
+            drivingFeatures.setIntervalsForTraversal(intervalsForTraversal);
+        }
+    }
+
+    protected abstract boolean isAcceptableFeatureType(Class<? extends Feature> featureType);
+
+    /**
+     * Implementation of variant-based traversal.
+     * Subclasses can override to provide their own behavior but default implementation should be suitable for most uses.
+     */
+    @Override
+    public void traverse() {
+        // Process each variant in the input stream.
+        StreamSupport.stream(drivingFeatures.spliterator(), false)
+                .forEach(feature -> {
+                    final SimpleInterval variantInterval = new SimpleInterval(feature);
+                    apply(feature,
+                            new ReadsContext(reads, variantInterval),
+                            new ReferenceContext(reference, variantInterval),
+                            new FeatureContext(features, variantInterval));
+                    progressMeter.update(feature);
+                });
+    }
+
+    /**
+     * Process an individual variant. Must be implemented by tool authors.
+     * In general, tool authors should simply stream their output from apply(), and maintain as little internal state
+     * as possible.
+     *
+     * @param feature Current variant being processed.
+     * @param readsContext Reads overlapping the current variant. Will be an empty, but non-null, context object
+     *                     if there is no backing source of reads data (in which case all queries on it will return
+     *                     an empty array/iterator)
+     * @param referenceContext Reference bases spanning the current variant. Will be an empty, but non-null, context object
+     *                         if there is no backing source of reference data (in which case all queries on it will return
+     *                         an empty array/iterator). Can request extra bases of context around the current read's interval
+     *                         by invoking {@link ReferenceContext#setWindow}
+     *                         on this object before calling {@link ReferenceContext#getBases}
+     * @param featureContext Features spanning the current variant. Will be an empty, but non-null, context object
+     *                       if there is no backing source of Feature data (in which case all queries on it will return an
+     *                       empty List).
+     */
+    public abstract void apply(final F feature, final ReadsContext readsContext, final ReferenceContext referenceContext, final FeatureContext featureContext );
+
+    /**
+     * Close the reads and reference data sources.
+     *
+     * Marked final so that tool authors don't override it. Tool authors should override onTraversalDone() instead.
+     */
+    @Override
+    protected final void onShutdown() {
```

Comment:
Add a `final` override of `onStartup()` as well that just calls `super.onStartup();`, to prevent tool authors from overriding it.